File: 3. Max Sum of Distinct SubArrays/SlidingWindow+HashSet/sliding-window.ts
const maxSumDistinctSubArray = (nums: number[], k: number): number => {
  let maxSum = 0,
    sum = 0;
  let left = 0;
  const windowSet = new Set<number>();

  for (let right = 0; right < nums.length; right++) {
    // Maintain Uniqueness
    while (windowSet.has(nums[right])) {
      windowSet.delete(nums[left]);
      sum -= nums[left];
      left++;
    }

    // Expand window
    windowSet.add(nums[right]);
    sum += nums[right];

    // If window size exceeds k, shrink from the left
    if (right - left + 1 > k) {
      windowSet.delete(nums[left]);
      sum -= nums[left];
      left++;
    }

    // If window size is exactly k, update the maxSum
    if (right - left + 1 === k) {
      maxSum = Math.max(maxSum, sum);
    }
  }
  return maxSum;
};

File: 3. Max Sum of Distinct SubArrays/SlidingWindow+HashSet/sliding-window.cpp
#include <iostream>
#include <vector>
#include <unordered_set>

using namespace std;

long long maximumSubarraySum(const vector<int>& nums, int k) {
  int maxSum = 0, sum = 0;
  int left = 0;
  unordered_set<int> windowSet;

  for (int right = 0; right < nums.size(); ++right) {
    while (windowSet.count(nums[right])) {
      windowSet.erase(nums[left]);
      sum -= nums[left];
      ++left;
    }

    windowSet.insert(nums[right]);
    sum += nums[right];

    if (right - left + 1 > k) {
      windowSet.erase(nums[left]);
      sum -= nums[left];
      ++left;
    }

    if (right - left + 1 == k) {
      maxSum = max(maxSum, sum);
    }
    return maxSum;
  }
}
